File: packages/ts/tracing/src/debug-delivery.ts
export const PRODUCTION_BASE_URL = "https://api.uselemma.ai";
export const INGEST_PATH = "/traces/ingest";
export const INGEST_STATUS_PATH = "/traces/ingest-status";
export const EXPECTED_INGEST_SUCCESS_STATUS = 201;

/** Poll interval for LEMMA_DEBUG_VERIFY ingest-status checks. */
export const INGEST_STATUS_POLL_INTERVAL_MS = 1_000;
/** Total wait budget for LEMMA_DEBUG_VERIFY ingest-status checks. */
export const INGEST_STATUS_POLL_TIMEOUT_MS = 15_000;

export type IngestStatus = "enqueued" | "ingested" | "ready" | "not_found";

const SUCCESS_INGEST_STATUSES = new Set<IngestStatus>([
  "enqueued",
  "ingested",
  "ready",
]);

export function isSuccessfulIngestStatus(
  status: IngestStatus | null | undefined,
): status is Exclude<IngestStatus, "not_found"> {
  return status != null && SUCCESS_INGEST_STATUSES.has(status);
}

export function parseIngestStatus(value: unknown): IngestStatus | null {
  if (
    value === "enqueued" ||
    value === "ingested" ||
    value === "ready" ||
    value === "not_found"
  ) {
    return value;
  }
  return null;
}

const UUID_REGEX =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function isValidProjectId(projectId: string): boolean {
  return UUID_REGEX.test(projectId);
}

export function apiKeySuffix(apiKey: string): string {
  if (apiKey.length <= 4) return apiKey;
  return `...${apiKey.slice(-4)}`;
}

export function buildConfigWarnings(
  baseUrl: string,
  projectId: string,
): string[] {
  const warnings: string[] = [];
  if (baseUrl !== PRODUCTION_BASE_URL) {
    warnings.push(`baseUrl is not production (${PRODUCTION_BASE_URL})`);
  }
  if (!isValidProjectId(projectId)) {
    warnings.push("projectId is not a valid UUID");
  }
  return warnings;
}

export function ingestFailureHint(status: number): string | undefined {
  switch (status) {
    case 401:
      return "check LEMMA_API_KEY";
    case 403:
      return "API key doesn't own this project_id";
    case 429:
      return "ingest rate limit exceeded; retry with backoff";
    case 404:
      return "baseUrl likely wrong (not Lemma API)";
    default:
      return undefined;
  }
}

export function pickResponseHeaders(
  headers: Headers,
): Record<string, string> {
  const picked: Record<string, string> = {};
  for (const name of ["cf-ray", "server", "date"]) {
    const value = headers.get(name);
    if (value) picked[name] = value;
  }
  return picked;
}

export function pickResponseHeadersFromRecord(
  headers: Record<string, string>,
): Record<string, string> {
  const lowered = Object.fromEntries(
    Object.entries(headers).map(([key, value]) => [key.toLowerCase(), value]),
  );
  const picked: Record<string, string> = {};
  for (const name of ["cf-ray", "server", "date"]) {
    const value = lowered[name];
    if (value) picked[name] = value;
  }
  return picked;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: packages/py/tracing/src/uselemma_tracing/debug_delivery.py
from __future__ import annotations

import re

PRODUCTION_BASE_URL = "https://api.uselemma.ai"
INGEST_PATH = "/traces/ingest"
INGEST_STATUS_PATH = "/traces/ingest-status"
EXPECTED_INGEST_SUCCESS_STATUS = 201

# Poll interval / budget for LEMMA_DEBUG_VERIFY ingest-status checks.
INGEST_STATUS_POLL_INTERVAL_SECONDS = 1.0
INGEST_STATUS_POLL_TIMEOUT_SECONDS = 15.0

IngestStatus = str  # "enqueued" | "ingested" | "ready" | "not_found"
_SUCCESS_INGEST_STATUSES = frozenset({"enqueued", "ingested", "ready"})


def is_successful_ingest_status(status: str | None) -> bool:
    return status in _SUCCESS_INGEST_STATUSES


def parse_ingest_status(value: object) -> str | None:
    if value in {"enqueued", "ingested", "ready", "not_found"}:
        return str(value)
    return None

_UUID_REGEX = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)


def is_valid_project_id(project_id: str) -> bool:
    return bool(_UUID_REGEX.fullmatch(project_id))


def api_key_suffix(api_key: str) -> str:
    if len(api_key) <= 4:
        return api_key
    return f"...{api_key[-4:]}"


def build_config_warnings(base_url: str, project_id: str) -> list[str]:
    warnings: list[str] = []
    if base_url != PRODUCTION_BASE_URL:
        warnings.append(f"baseUrl is not production ({PRODUCTION_BASE_URL})")
    if not is_valid_project_id(project_id):
        warnings.append("projectId is not a valid UUID")
    return warnings


def ingest_failure_hint(status: int) -> str | None:
    hints = {
        401: "check LEMMA_API_KEY",
        403: "API key doesn't own this project_id",
        429: "ingest rate limit exceeded; retry with backoff",
        404: "baseUrl likely wrong (not Lemma API)",
    }
    return hints.get(status)


def pick_response_headers(headers: dict[str, str]) -> dict[str, str]:
    picked: dict[str, str] = {}
    lowered = {key.lower(): value for key, value in headers.items()}
    for name in ("cf-ray", "server", "date"):
        value = lowered.get(name)
        if value:
            picked[name] = value
    return picked
